include/beast/websocket/impl/read_frame_op.ipp

```diff
@@ -129,373 +130,415 @@ template<class NextLayer>
 template<class Buffers, class Handler>
 void
 stream<NextLayer>::read_frame_op<Buffers, Handler>::
+operator()(error_code ec, std::size_t bytes_transferred)
+{
+    auto& d = *d_;
+    if(ec)
+        d.ws.failed_ = true;
+    (*this)(ec, bytes_transferred, true);
+}
+
+template<class NextLayer>
+template<class Buffers, class Handler>
+void
+stream<NextLayer>::read_frame_op<Buffers, Handler>::
 operator()(error_code ec,std::size_t bytes_transferred, bool again)
 {
+    enum
+    {
+        do_start = 0,
+        do_read_payload = 1,
+        do_frame_done = 3,
```

Comment:
I like how the state machine turned out with the enum and the `case state+1:` idiom. One idea is to reserve some low-order bits for this purpose. i.e all enums are numbered sequentially with the pattern `do_start = 0 << 2, do_read_payload = 1 << 2,` etc. That makes it easy to audit code (if I ever see `state+4` I know it's an error).